datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -643,9 +643,8 @@ fn dictionary_coercion(
             Dictionary(_lhs_index_type, lhs_value_type),
             Dictionary(_rhs_index_type, rhs_value_type),
         ) => comparison_coercion(lhs_value_type, rhs_value_type),
-        (d @ Dictionary(_, value_type), other_type)
-        | (other_type, d @ Dictionary(_, value_type))
-            if preserve_dictionaries && value_type.as_ref() == other_type =>
+        (d @ Dictionary(_, _), other_type) | (other_type, d @ Dictionary(_, _))
+            if preserve_dictionaries && can_cast_types(other_type, d) =>
```

Comment:
If I understand correctly, we should check whether `value_type` and `other_type` are coercible. If yes return the new dict with new value type

`Dict(k, i8) and i64 -> Dict(k, i64)`

Instead of `Some(d.clone())` we should create a new dict with new value type.

And, I think we should use the coerce logic in `comparison_coercion` not `can_cast_types`, because the logic is not totally the same.

```rust
        | (other_type, d @ Dictionary(key_type, value_type))
            if preserve_dictionaries && comparison_coercion(value_type.as_ref(), other_type).is_some() =>
        {
            let new_value = comparison_coercion(&value_type, other_type).unwrap();
            Some(DataType::Dictionary(key_type.clone(), Box::new(new_value)))
        }
```

Reply:
@jayzhan211 Thanks for the tip.  I initially avoiding calling `comparison_coercion` because the way I was doing it would cause an infinite recursion, but calling `comparison_coercion` on the inner value type and then wrapping in a new `Dictionary` as you suggested solves that issue. 

I was hoping this would fix the test failure, but unfortunately it does not so I will need to look into reworking the type coercion logic for `VariadicEqual` functions